File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverse= function(start,end){
    let curr=start,prev=null;
    while(curr!=end){
        let temp = curr.next;
        curr.next=prev;
        prev=curr;
        curr=temp;
    }
    return prev;
}
var reverseKGroup = function(head, k) {
    if(head==null || head.next==null || k==1) return head;
    let start=head,end=head;
    for(let i=0;i<k;i++){
        if(end==null)return head;
        end=end.next;
    }
    let newNode = reverse(start,end);
    start.next=reverseKGroup(end,k);
    return newNode;

};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* start,ListNode*end){
        ListNode* curr=start;
        ListNode* prev=NULL;
        while(curr!=end){
            ListNode* temp=curr->next;
            curr->next=prev;
            prev=curr;
            curr=temp;
        }
        return prev;

    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head==NULL || head->next==NULL || k==1) return head;
        ListNode* start=head;
        ListNode* end=head;
        for(int i=0;i<k;i++){
            if(end==NULL) return head;
            end=end->next;
        }
        ListNode* newNode = reverse(start,end);
        start->next=reverseKGroup(end,k);
        return newNode;
    }
};